src/lib/snark_work_lib/partitioned.ml

```diff
@@ -0,0 +1,534 @@
+    let proof = Ledger_proof.Cached.write_proof_to_disk ~proof_cache_db proof in
+    { proof; elapsed }
+end
+
+let construct_selector_result
+    ~(instances : (Selector.Single.Spec.t * Proof_with_metric.t) One_or_two.t)
+    ~fee ~prover =
+  let proofs = One_or_two.map ~f:(fun (_, { proof; _ }) -> proof) instances in
+  let metrics =
+    One_or_two.map
+      ~f:(fun (single, { elapsed; _ }) ->
+        let tag =
+          match single with
+          | Work.Single.Spec.Transition (_, _) ->
+              `Transition
+          | Work.Single.Spec.Merge (_, _, _) ->
+              `Merge
+        in
+        (elapsed, tag) )
+      instances
+  in
+
+  let instances = One_or_two.map ~f:(fun (single, _) -> single) instances in
+  ({ proofs; metrics; spec = { instances; fee }; prover } : Selector.Result.t)
+
+module Result = struct
+  [%%versioned
+  module Stable = struct
+    [@@@no_toplevel_latest_type]
+
+    module V1 = struct
+      type t =
+        { (* Throw everything inside the spec to ensure proofs, metrics have correct shape *)
+          data : Proof_with_metric.Stable.V1.t Spec.Poly.Stable.V1.t
+        ; prover : Signature_lib.Public_key.Compressed.Stable.V1.t
+        }
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t =
+    { data : Proof_with_metric.t Spec.Poly.t
+    ; prover : Signature_lib.Public_key.Compressed.t
+    }
+
+  let read_all_proofs_from_disk ({ data; prover } : t) : Stable.Latest.t =
+    let data =
+      Spec.Poly.(
+        map_metric ~f:Proof_with_metric.read_all_proofs_from_disk data
+        |> read_all_proofs_from_disk)
+    in
+
+    { data; prover }
+
+  let write_all_proofs_to_disk ~proof_cache_db
+      ({ data; prover } : Stable.Latest.t) : t =
+    let data =
+      Spec.Poly.(
+        write_all_proofs_to_disk ~proof_cache_db data
+        |> map_metric
+             ~f:(Proof_with_metric.write_all_proofs_to_disk ~proof_cache_db))
+    in
+    { data; prover }
+
+  let to_spec ({ data; _ } : t) : Spec.t =
+    match data with
+    | Spec.Poly.Single { single_spec; pairing; common; _ } ->
+        Spec.Poly.Single { single_spec; pairing; common; metric = () }
+    | Spec.Poly.Sub_zkapp_command { spec; _ } ->
+        Spec.Poly.Sub_zkapp_command { spec; metric = () }
+    | Spec.Poly.Old { instances; common } ->
+        Spec.Poly.Old
+          { instances =
+              One_or_two.map ~f:(fun (spec, _) -> (spec, ())) instances
+          ; common
+          }
+
+  let of_selector_result ~issued_since_unix_epoch
+      ({ proofs; metrics; spec = { instances; fee } as spec; prover } :
+        Selector.Result.t ) : t Or_error.t =
+    let issued_since_unix_epoch =
+      match issued_since_unix_epoch with
+      | None ->
+          let spec_str =
+            Selector.Spec.read_all_proofs_from_disk spec
+            |> Selector.Spec.Stable.V1.to_yojson |> Yojson.Safe.to_string
+          in
+          printf "No issued time provided, assuming now for %s" spec_str ;
```

Comment:
Don't we want to log this instead of using the stdin?

Reply:
I removed the optional and make it mandatary